src/wikidata.py

```diff
@@ -0,0 +1,224 @@
+            }}
+            LIMIT {current_page_size}
+            OFFSET {offset}
+            """
+
+        try:
+            response = self.session.get(
+                self.WIKIDATA_SPARQL_BASE_URL,
+                params={"query": query, "format": "json"},
+            )
+            response.raise_for_status()
+
+            results = response.json()["results"]["bindings"]
+            qids: list[WikibaseID] = []
+            for result in results:
+                item_id = result["item"]["value"].split("/")[-1]
+                qids.append(WikibaseID(item_id))
+
+            logger.debug(
+                f"Found {len(qids)} entities related to {entity_id} through property {property_id} "
+                f"(inverse={inverse}, offset={offset}, page_size={page_size})"
+            )
+
+            # Check whether we've got a full page of results and if we're still under the
+            # user-specified limit. If so, fetch the next page of results
+            got_a_full_page_of_results = len(qids) == current_page_size
+            still_under_the_limit = limit is None or (offset + len(qids)) < limit
+
+            if got_a_full_page_of_results and still_under_the_limit:
+                next_page = self.get_property_values(
+                    property_id,
+                    entity_id,
+                    inverse,
+                    limit,
+                    page_size,
+                    offset + len(qids),
+                )
+                qids.extend(next_page)
+
+            return qids
+
+        except httpx.HTTPError as e:
+            logger.error(
+                f"HTTP error during SPARQL query for {entity_id} with property {property_id}: {e}"
+            )
+            raise
+        except KeyError as e:
+            logger.error(
+                f"Unexpected response format for {entity_id} with property {property_id}: {e}"
+            )
+            raise ValueError(
+                f"Unexpected response format from Wikidata SPARQL endpoint: {e}"
+            )
+        except Exception as e:
+            logger.error(
+                f"Error during SPARQL query for {entity_id} with property {property_id}: {e}"
+            )
+            raise
+
+    def get_instances_of(
+        self, entity_id: WikibaseID, limit: Optional[int] = None
+    ) -> list[WikibaseID]:
+        """
+        Get all instances of a given Wikidata entity.
+
+        :param WikibaseID entity_id: The Wikidata ID to find instances of
+        :param Optional[int] limit: Maximum number of results to return
+        :return list[WikibaseID]: A list of Wikidata IDs that are instances of the entity
+        """
+        return self.get_property_values(
+            self.INSTANCE_OF_PROPERTY, entity_id, limit=limit
+        )
+
+    def get_parent_entities(
+        self, entity_id: WikibaseID, limit: Optional[int] = None
+    ) -> list[WikibaseID]:
+        """
+        Get all parent entities of a given Wikidata entity.
+
```

Comment:
it might be worth specifying that this is using the 'instance of' property here. When i initially read this i thought it could be related to 'subclass of' (which is probably my misunderstanding of ontologies/hierarchies, but also then likely others'!)

Reply:
yeah, i really struggled with the naming of this method! 

i usually love giving painfully long, unambiguous names to methods/variables/tests, but `get_items_of_which_the_specified_item_is_said_to_be_an_instance()` felt like a stretch, even for me 😅 

docstring definitely has room for improvement though 🙂

Reply:
renamed to `get_instance_of_values()`. even if it's a bit harder to understand what the method does at a glance, i think it's also harder to _misunderstand_ at a glance, hopefully nudging users to read the docstring 🤞